README.md
# badgeit

> 🎫 if you like it, put a badge on it

[![Build Status](https://travis-ci.org/gabrielcsapo/badgeit.svg?branch=master)](https://travis-ci.org/gabrielcsapo/badgeit)
[![Dependency Status](https://starbuck.gabrielcsapo.com/badge/github/gabrielcsapo/badgeit/status.svg)](https://starbuck.gabrielcsapo.com/github/gabrielcsapo/badgeit)
[![devDependency Status](https://starbuck.gabrielcsapo.com/badge/github/gabrielcsapo/badgeit/dev-status.svg)](https://starbuck.gabrielcsapo.com/github/gabrielcsapo/badgeit#info=devDependencies)
[![Coverage Status](https://lcov-server.gabrielcsapo.com/badge/github%2Ecom/gabrielcsapo/badgeit.svg)](https://lcov-server.gabrielcsapo.com/coverage/github%2Ecom/gabrielcsapo/badgeit)
[![npm](https://img.shields.io/npm/dt/badgeit.svg?maxAge=2592000)]()
[![npm](https://img.shields.io/npm/dm/badgeit.svg?maxAge=2592000)]()

## Installation

```
npm install --save badgeit
```

## Usage

```javascript
const badgeit = require('badgeit');

const badge = await badgeit({ text: ['Hello', 'World'] });
```


CHANGELOG.md
# 0.1.1 (05/14/2019)

- adds npm path to publish fonts folder

# 0.1.0 (11/26/2018)

- updates `babel-core` -> `@babel/core`
- moves from `eslint` -> `standard`
- updates `opentype.js@0.7.3` -> `opentype.js@0.11.0`

# 0.0.2 (11/07/2017)

- fixes usability with node

# 0.0.1 (11/07/2017)

- initial functionality working


test/index.js
const fs = require('fs')
const path = require('path')
const test = require('tape')

const badge = require('../index')

function clean (string) {
  return string.replace(/ /g, '').replace(/\n/g, '')
}

test('badgeit', (t) => {
  t.plan(5)

  t.test('should work with nothing passed in', async (t) => {
    t.plan(1)

    try {
      t.equal(clean(await badge()), clean(fs.readFileSync(path.resolve(__dirname, 'fixtures', 'basic.svg')).toString('utf8')))
    } catch (ex) {
      t.fail(ex)
    }
  })

  t.test('should work with overrides (text)', async (t) => {
    t.plan(1)

    try {
      t.equal(clean(await badge({ text: ['hello world', '100%'] })), clean(fs.readFileSync(path.resolve(__dirname, 'fixtures', 'override_text.svg')).toString('utf8')))
    } catch (ex) {
      t.fail(ex)
    }
  })

  t.test('should work with overrides (colors)', async (t) => {
    t.plan(1)

    try {
      t.equal(clean(await badge({ text: ['coverage', '100%'], colors: { left: '#eb46ff', right: '#cff2c2', font: '#ffffff' } })), clean(fs.readFileSync(path.resolve(__dirname, 'fixtures', 'override_color.svg')).toString('utf8')))
    } catch (ex) {
      t.fail(ex)
    }
  })

  t.test('should throw if font path doesn\'t exist', async (t) => {
    t.plan(2)

    try {
      await badge({ fontPath: '/nowhere/land' })
      t.fail('should have thrown error')
    } catch (ex) {
      t.ok(ex instanceof Error)
      t.equal(ex.message, 'font could not be loaded')
    }
  })

  t.test('should throw if text doesn\'t contain two string elements', async (t) => {
    t.plan(2)

    try {
      await badge({ text: ['coverage'] })
      t.fail('should have thrown error')
    } catch (ex) {
      t.ok(ex instanceof Error)
      t.equal(ex.message, 'text should contain an array of two string elements')
    }
  })
})


tryitout.js
const { name, description } = require('./package.json');

module.exports = {
	title: name,
	description: description,
	nav: {
		Source: 'https://github.com/gabrielcsapo/badgeit',
	},
	body: [{
		type: 'code',
		title: 'Simple example',
		subtitle: 'Passing an array with two values changes the text',
		value: `
        const badge = require('badgeit');
        badge({ text: ['coverage', '72%' ], colors: { right: '#caa300' }})
          .then((html) => console.html(html))
          .catch((ex) => console.log(ex.toString()));
      `
	}, {
		type: 'text',
		value: `
> The following are the defaults and their types

\`\`\`
{
	fontPath: "/OpenSans-Bold" // String
	text: ['Hello', 'World'], // Array<String, String>
	fontSize: 11, // Number
	colors: {
		left: '#555', // String
		right: '#4c1', // String
		font: '#fff', // String
		shadow: '#010101' // String
	},
	paddingX: 6, // Number
	paddingY: 6, // Number
	offsetX: 0 // Number
}
\`\`\`
		`
	}],
	output: './docs',
	externals: [
		'./dist/badgeit.min.js'
	]
};


index.js
const fs = require('fs')
const path = require('path')
const opentype = require('opentype.js')
const defaults = require('lodash.defaultsdeep')
const defaultFont = fs.readFileSync(path.resolve(__dirname, 'fonts/OpenSans-Bold.ttf'))

function text (message, options) {
  const { font, offsetX, paddingX, paddingY, fontSize } = options

  var scaleFactor = fontSize / font.unitsPerEm
  var heightInUnits = font.charToGlyph('E')
  heightInUnits.getPath()
  heightInUnits = heightInUnits.yMax
  var heightInPixels = Math.round(heightInUnits * scaleFactor)

  var widthInUnits = font
    .stringToGlyphs(message)
    .reduce((w, glyph) => w + glyph.advanceWidth, 0)
  var widthInPixels = Math.round(widthInUnits * scaleFactor)
  var textPath = font
    .getPath(message, paddingX + offsetX, heightInPixels + paddingY, fontSize)
    .toSVG(1)
    .replace(/\.0 /g, ' ') // Pack it tighter by removing unused decimal places

  return {
    width: (paddingX * 2) + widthInPixels,
    height: (paddingY * 2) + heightInPixels,
    textPath: textPath
  }
}

module.exports = function badgeit (options = {}) {
  return new Promise(function (resolve, reject) {
    let font

    if (options.text && options.text.length !== 2) {
      return reject(new Error('text should contain an array of two string elements'))
    }

    if (options.fontPath) {
      try {
        font = opentype.loadSync(options.fontPath)
      } catch (ex) {
        return reject(new Error('font could not be loaded'))
      }
    } else {
      const fontBuffer = defaultFont.buffer.slice(defaultFont.byteOffset, defaultFont.byteOffset + defaultFont.byteLength)
      font = opentype.parse(fontBuffer)
    }

    const values = defaults(options, {
      font,
      text: ['Hello', 'World'],
      fontSize: 11,
      colors: {
        left: '#555',
        right: '#4c1',
        font: '#fff',
        shadow: '#010101'
      },
      paddingX: 6,
      paddingY: 6,
      offsetX: 0
    })

    const rightText = text(values.text[0], values)
    const leftText = text(values.text[1], Object.assign(values, {
      offsetX: rightText.width
    }))

    const width = rightText.width + leftText.width
    const height = rightText.height

    return resolve(`<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${width}" height="${height}">
          <defs>
              <linearGradient id="glow" x2="0" y2="100%">
                  <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
                  <stop offset="1" stop-opacity=".1"/>
              </linearGradient>
              <mask id="mask">
                  <rect width="${width}" height="${height}" rx="3" fill="#fff"/>
              </mask>
              <g id="text">
                  ${rightText.textPath}
                  ${leftText.textPath}
              </g>
          </defs>

          <g mask="url(#mask)">
              <rect fill="${values.colors.left}" x="0" y="0" width="${rightText.width}" height="${height}"/>
              <rect fill="${values.colors.right}" x="${rightText.width}" y="0" width="${width}" height="${height}"/>
              <rect fill="url(#glow)" x="0" y="0" width="${width}" height="${height}"/>
          </g>
          <g fill="${values.colors.font}">
              <use x="0" y="1" fill="${values.colors.shadow}" fill-opacity=".3" xlink:href="#text"/>
              <use x="0" y="0" xlink:href="#text"/>
          </g>
      </svg>`)
  })
}
